server/src/main/java/org/elasticsearch/index/mapper/FieldTypeLookup.java

```diff
@@ -65,18 +68,39 @@ final class FieldTypeLookup {
             for (String targetField : fieldMapper.copyTo().copyToFields()) {
                 Set<String> sourcePath = fieldToCopiedFields.get(targetField);
                 if (sourcePath == null) {
+                    // TODO: Any concerns about copy field order due to set usage?
                     Set<String> copiedFields = new HashSet<>();
                     copiedFields.add(targetField);
                     fieldToCopiedFields.put(targetField, copiedFields);
                 }
                 fieldToCopiedFields.get(targetField).add(fieldName);
             }
             if (fieldType instanceof InferenceModelFieldType inferenceModelFieldType) {
-                String inferenceModel = inferenceModelFieldType.getInferenceModel();
-                if (inferenceModel != null) {
-                    Set<String> fields = fieldsForModels.computeIfAbsent(inferenceModel, v -> new HashSet<>());
-                    fields.add(fieldName);
-                }
+                // Add this field type to a list of ones we will handle in a second pass, after we have processed the full
+                // multi-field/copy_to context
+                inferenceModelFieldTypes.add(inferenceModelFieldType);
+            }
+        }
+
+        for (InferenceModelFieldType fieldType : inferenceModelFieldTypes) {
+            String fieldName = fieldType.name();
+            String inferenceModel = fieldType.getInferenceModel();
+            if (inferenceModel == null) {
+                // TODO: Should we log when this happens?
```

Comment:
Alternately should this be a case when we return an error?

Reply:
Probably yes. `InferenceModelFieldType` implementations should validate on their own that the inference model is specified, so in theory we should never hit this condition. If we do, it's an indicator of an error in the `InferenceModelFieldType` implementation. Given that, maybe `IllegalStateException` is appropriate here? If that doesn't fit, maybe `ElasticsearchException`?